systemtest/edu/washington/escience/myria/systemtest/UploadDownloadS3Test.java

```diff
@@ -0,0 +1,98 @@
+    String filePath = Paths.get("testdata", "filescan", "simple_two_col_int_to_hash.txt").toString();
+    DataSource relationSource = new FileSource(filePath);
+    RelationKey relationKeyUpload = RelationKey.of("public", "adhoc", "upload");
+    Schema relationSchema = Schema.ofFields("x", Type.INT_TYPE, "y", Type.INT_TYPE);
+
+    JsonAPIUtils.ingestData("localhost", masterDaemonPort, ingest(relationKeyUpload, relationSchema, relationSource,
+        ' ', new RoundRobinPartitionFunction(workerIDs.length)));
+
+    /* File to upload and download */
+    String fileName = String.format("s3://myria-test/test.txt");
+
+    /* Construct the query and upload data */
+    ExchangePairID serverReceiveID = ExchangePairID.newID();
+    DbQueryScan workerScan = new DbQueryScan(relationKeyUpload, relationSchema);
+    CollectProducer workerProduce = new CollectProducer(workerScan, serverReceiveID, MASTER_ID);
+
+    HashMap<Integer, RootOperator[]> workerPlans = new HashMap<Integer, RootOperator[]>();
+    for (int workerID : workerIDs) {
+      workerPlans.put(workerID, new RootOperator[] { workerProduce });
+    }
+    CollectConsumer serverCollect = new CollectConsumer(relationSchema, serverReceiveID, workerIDs);
+    InMemoryOrderBy sortOperator = new InMemoryOrderBy(serverCollect, new int[] { 1 }, new boolean[] { true });
+    DataSink dataSink = new UriSink(fileName);
+    DataOutput masterRoot = new DataOutput(sortOperator, new CsvTupleWriter(), dataSink);
+    server.submitQueryPlan(masterRoot, workerPlans).get();
+
+    /* Read the data back in from S3 and shuffle to one worker */
+    RelationKey relationKeyDownload = RelationKey.of("public", "adhoc", "download");
+    DataSource relationSourceS3 = new UriSource(fileName);
+
+    ExchangePairID workerReceiveID = ExchangePairID.newID();
+    FileScan serverFileScan = new FileScan(relationSourceS3, relationSchema, ',', null, null, 1);
+    GenericShuffleProducer serverProduce =
+        new GenericShuffleProducer(serverFileScan, workerReceiveID, workerIDs, new SingleFieldHashPartitionFunction(
+            workerIDs.length, 0));
+    GenericShuffleConsumer workerConsumer =
+        new GenericShuffleConsumer(relationSchema, workerReceiveID, new int[] { MASTER_ID });
+    DbInsert workerInsert = new DbInsert(workerConsumer, relationKeyDownload, true);
+    HashMap<Integer, RootOperator[]> workerPlansInsert = new HashMap<Integer, RootOperator[]>();
+    for (int workerID : workerIDs) {
+      workerPlansInsert.put(workerID, new RootOperator[] { workerInsert });
+    }
+    server.submitQueryPlan(serverProduce, workerPlansInsert).get();
+
+    String dstData =
+        JsonAPIUtils.download("localhost", masterDaemonPort, relationKeyDownload.getUserName(), relationKeyDownload
+            .getProgramName(), relationKeyDownload.getRelationName(), "json");
+
+    String srcData =
+        "[{\"x\":1,\"y\":2},{\"x\":1,\"y\":2},{\"x\":1,\"y\":4},{\"x\":1,\"y\":4},{\"x\":1,\"y\":6},{\"x\":1,\"y\":6}]";
```

Comment:
Not a big fan of storing the JSON form of the test data independently of its source file, but if constructing the JSON string in the right order from the source data is ridiculously complicated then I guess it's OK.